Source/missiles.cpp

```diff
@@ -590,6 +590,90 @@ void CheckMissileCol(Missile &missile, int mindam, int maxdam, bool shift, Point
 		PlaySfxLoc(MissilesData[missile._mitype].miSFX, missile.position.tile);
 }
 
+/**
+ * @brief Could the missile collide with solid objects? (like walls or closed doors)
+ */
+bool CouldMissileCollideWithSolidObject(Point tile)
+{
+	int oid = dObject[tile.x][tile.y];
+	if (oid != 0) {
+		oid = oid > 0 ? oid - 1 : -(oid + 1);
+		if (!Objects[oid]._oMissFlag)
+			return true;
+	}
+	return nMissileTable[dPiece[tile.x][tile.y]];
+}
+
+void MoveMissileAndCheckMissileCol(Missile &missile, int mindam, int maxdam, bool ignoreStart, bool ifCollidesDontMoveToHitTile)
+{
+	Point prevTile = missile.position.tile;
+	missile.position.traveled += missile.position.velocity;
+	UpdateMissilePos(missile);
+	int walkingDistance = prevTile.WalkingDistance(missile.position.tile);
+
+	int16_t tileTargetIdentifer = dMonster[missile.position.tile.x][missile.position.tile.y] - dPlayer[missile.position.tile.x][missile.position.tile.y];

```

Comment:
Given the way dMonster and dPlayer contain negated indexes for moving entities I think using bitwise combinations would be more straightforward than subtraction here.

XOR should be fine, could also shift the value in dMonster left 4-8 bits and mask the player index onto the low bits to avoid collisions (which is already pretty unlikely).
```suggestion
	int16_t tileTargetIdentifer = dMonster[missile.position.tile.x][missile.position.tile.y] ^ dPlayer[missile.position.tile.x][missile.position.tile.y];
```

Reply:
Rather than unlikely, I think collisions are impossible. Can players and monsters occupy the same tile simultaneously?

Reply:
If the indexes are negated for moving entities, doesn't that mean we might still check collisions twice for the same entity in the same tile if the entity starts or stops moving while the missile travels through that tile? Perhaps we should use `abs()`.

Reply:
> Rather than unlikely, I think collisions are impossible. Can players and monsters occupy the same tile simultaneously?
> 
> EDIT: I suppose maybe you meant hash collisions. 🤦

I did mean hash collisions :D, however hash collisions are only an issue if a player occupies the same tile as the golem they control, which I don't think is possible even if both entities are moving? Given the point you bring up in the second comment about hitting a moving entity multiple times using the first non-zero absolute value sounds fine.

Reply:
> I did mean hash collisions :D, however hash collisions are only an issue if a player occupies the same tile as the golem they control, which I don't think is possible even if both entities are moving?

I have bad news.

https://user-images.githubusercontent.com/9203145/132784190-43ff7aec-6fd4-44ef-8e80-75481b2cb77f.mp4

Reply:
While scanning through the DSF Buglist just now, I found this text on the subject of sharing tiles.

> **SHARING SQUARES**
> 
> When a character teleports/phases onto a square that a monster was moving in to, they will overlap. The character will have to move off the square to hit the monster. The monster, however, can attack the same square it's standing on.
> 
> If two characters wind up occupying the same square, which can happen rarely when casting Teleport or when Resurrecting while standing over a body, the player who does not move will become invisible to both players, until he/she moves. Further, the player first on the internal list of players may be the first targeted and attacked if a monster approaches.
> 
> The player can also share a square with a Golem.

Reply:
A player can also share a tile with a monster by spawning on that tile while a monster occupies it. The function for spawning players on a floor (such as by taking the stairs down) will put a player on a tile occupied by a monster happily. When this happens, the player cannot hit the monster until they move, but the monster is free to melee the player to death. Fun times. (if you attack the monster, your attack goes straight down on screen)

Reply:
> I have bad news.

> A player can also share a tile with a monster by spawning on that tile while a monster occupies it.

I think it is a bad idea to base new logic on what are clearly existing bugs. Instead, we should fix the bugs and prevent the situation from ever happening.

Reply:
> I think it is a bad idea to base new logic on what are clearly existing bugs. Instead, we should fix the bugs and prevent the situation from ever happening.

On that note, I think the hash collisions are pretty much a non-issue. Even in an imagined future state of the game where any monster could share an ID with any player, it would be extremely difficult to create a situation where the hash collision actually made a difference to this logic. Furthermore, the impact of the hash collision is negligible.

Reply:
I'm also not sure if we should care much about it.
For me this are also bugs that should be fixed and @qndel was fast and fixed the first one already. 👍 
Do we have a issue for the other player spawning case?

But if wanted we could improve it now or in a later pr.
We could for example add two fields to `Missile`, one for player and one for monster.

Reply:
I don't think there's an issue for players and monsters occupying the same tiles. Though, now I'm curious about other ways to possibly do it.... if I get some time this weekend, I'll try to get some save files and confirm things and get some issues open.

Reply:
I still think that absolute value is a good idea, but the impact is similarly negligible so I'll approve the PR regardless.

Reply:
I changed it to suggested `^`, it also doesn't hurt. 🙂